### PR TITLE
Improve include checker.

### PR DESCRIPTION
**Pull request by tammoippen**
The script `extras/include_checker.py` suggests an include order according to the [coding guidelines](https://nest.github.io/nest-simulator/coding_guidelines_c++#order-of-includes). This PR gives a warning for excluded files and recursively runs through directories.

### CODE REVIEW

File: extras/include_checker.py [hunk 1]
```diff
@@ -145,6 +145,7 @@ def print_includes(includes):
 
 def process_source(path, f, all_header, print_suggestion):
     if f in excludes_files:
+        print("Please do not change the includes in '" + f + "'.")
```
Comment (jougs) [inline extras/include_checker.py]: Please re-phrase: 
```python
print("Not checking file " + f +" as it is in the exclude list. Please do not change the order of includes.")
```

Also, please add a comment to line 27 explaining why files are in the exclude list. For a possible sentence on this, see https://github.com/nest/nest-simulator/pull/280#discussion_r60908734

**Comment by jougs** [comment]
See my comment above. Otherwise this is very nice. So :+1: once my comment is addressed. I think we can waive a second review as this is only a helper script.

File: extras/include_checker.py [hunk 2]
```diff
@@ -145,6 +147,7 @@ def print_includes(includes):
 
 def process_source(path, f, all_header, print_suggestion):
```
Comment (heplesser) [inline extras/include_checker.py]: Should be `all_headers` (plural) throughout.

File: extras/include_checker.py [hunk 3]
```diff
@@ -84,7 +86,7 @@ def to_string(self):
 
 def all_includes(path):
     result = {}
-    dirs = [d for d in next(os.walk(path))[1] if d[0] != '.' and not d in excludes ]
+    dirs = [d for d in next(os.walk(path))[1] if d[0] != '.' ]
     for d in dirs:
         for root, dirs, files in os.walk(path + "/" + d):
```
Comment (heplesser) [inline extras/include_checker.py]: Use `os.path.join()` to compose paths.

File: extras/include_checker.py [hunk 4]
```diff
@@ -145,6 +147,7 @@ def print_includes(includes):
 
 def process_source(path, f, all_header, print_suggestion):
     if f in excludes_files:
+        print("Not checking file " + f + " as it is in the exclude list. Please do not change the order of includes.")
         return 0
     includes = get_includes_from(path + "/" + f, all_header)
```
Comment (heplesser) [inline extras/include_checker.py]: Use `os.path.join()` to compose paths.

File: extras/include_checker.py [hunk 5]
```diff
@@ -164,21 +167,21 @@ def process_source(path, f, all_header, print_suggestion):
...
-    dirs = [d for d in next(os.walk(path))[1] if d[0] != '.' and not d in excludes ]
-    for d in dirs:
-        for root, dirs, files in os.walk(path + "/" + d):
-            for f in files:
-                if re.search("\.h$|\.hpp$|\.c$|\.cc|\.cpp$",f):
-                    # valid source file
-                    count += process_source(root, f, all_header, print_suggestion)
+    for root, dirs, files in os.walk(path):
+        for f in files:
+            if re.search("\.h$|\.hpp$|\.c$|\.cc|\.cpp$",f):
```
Comment (heplesser) [inline extras/include_checker.py]: Space missing before `f`.

File: extras/include_checker.py [hunk 6]
```diff
@@ -164,21 +167,21 @@ def process_source(path, f, all_header, print_suggestion):
...
+                # valid source file
```
Comment (heplesser) [inline extras/include_checker.py]: Indentation of this comment?

File: extras/include_checker.py [hunk 7]
```diff
@@ -164,21 +167,21 @@ def process_source(path, f, all_header, print_suggestion):
...
+                count += process_source(root, f, all_header, print_suggestion)
+        for d in dirs:
+            count += process_all_sources(root + "/" + d, print_suggestion)
```
Comment (heplesser) [inline extras/include_checker.py]: Use `os.path.join()` to compose paths.

**Comment by heplesser** [comment]
I think it would be very helpful to have a docstring explaining what this file is actually doing. I also think some lines might be a little long---have you run a pep8 check?

**Comment by heplesser** [comment]
Looks good now, merging.

**Pull request merged**